sputnikdao-factory2/src/factory_manager.rs

```diff
@@ -0,0 +1,167 @@
+    pub fn delete_contract(&self, code_hash: Base58CryptoHash) {
+        let code_hash: CryptoHash = code_hash.into();
+        env::storage_remove(&code_hash);
+    }
+
+    /// Get code for given hash.
+    pub fn get_code(&self, code_hash: Base58CryptoHash) {
+        let code_hash: CryptoHash = code_hash.into();
+        unsafe {
+            // Check that such contract exists.
+            assert_eq!(
+                sys::storage_has_key(code_hash.len() as _, code_hash.as_ptr() as _),
+                1,
+                "Contract doesn't exist"
+            );
+            // Load the hash from storage.
+            sys::storage_read(code_hash.len() as _, code_hash.as_ptr() as _, 0);
+            // Return as value.
+            sys::value_return(u64::MAX as _, 0 as _);
+        }
+    }
+
+    /// Forces update on the given contract.
+    /// Contract must support update by factory for this via permission check.
+    pub fn update_contract(
+        &self,
+        account_id: AccountId,
+        code_hash: Base58CryptoHash,
+        method_name: &str,
+    ) {
+        let code_hash: CryptoHash = code_hash.into();
+        let account_id = account_id.as_bytes().to_vec();
+        unsafe {
+            // Check that such contract exists.
+            assert_eq!(
+                sys::storage_has_key(code_hash.len() as _, code_hash.as_ptr() as _),
+                1,
+                "Contract doesn't exist"
+            );
+            // Load the hash from storage.
+            sys::storage_read(code_hash.len() as _, code_hash.as_ptr() as _, 0);
+            // Create a promise toward given account.
+            let promise_id =
+                sys::promise_batch_create(account_id.len() as _, account_id.as_ptr() as _);
+            // Call `update` method, which should also handle migrations.
+            sys::promise_batch_action_function_call(
+                promise_id,
+                method_name.len() as _,
+                method_name.as_ptr() as _,
+                u64::MAX as _,
+                0,
+                &NO_DEPOSIT as *const u128 as _,
+                (env::prepaid_gas() - env::used_gas() - GAS_UPDATE_LEFTOVER).0,
+            );
+            sys::promise_return(promise_id);
+        }
+    }
+
+    /// Create given contract with args and callback factory.
+    pub fn create_contract(
+        &self,
+        code_hash: Base58CryptoHash,
+        account_id: AccountId,
+        new_method: &str,
+        args: &[u8],
+        callback_method: &str,
+        callback_args: &[u8],
+    ) {
+        let code_hash: CryptoHash = code_hash.into();
+        let attached_deposit = env::attached_deposit();
+        let factory_account_id = env::current_account_id().as_bytes().to_vec();
+        let account_id = account_id.as_bytes().to_vec();
+        unsafe {
+            // Check that such contract exists.
+            assert_eq!(
+                sys::storage_has_key(code_hash.len() as _, code_hash.as_ptr() as _),
```

Comment:
In terms of readability, wouldn't be simpler to use the `env` wrappers? Ex:  `env::storage_has_key`, `storage_read` etc

Reply:
correct on `storage_has_key`, but `env::storage_read` doesn't allow to use registers.